doc/design/upgrade.md

```diff
@@ -125,15 +135,114 @@ Here is the upgrade workflow
+
+- Appliance guestinfo, which is used to persist VCH appliance configuration
+- Container guestinfo, which is used to persist container configuration
+- KeyValue Store, which is used to persist image metadata and network portlayer data
+- Appliance log files
+- vSphere object management logic change
+
+### Guestinfo Migration
+
+- One VirtualContainerHost structure per one version, this will be used by extraconfig package to read back old vch's configuration
+- Write plugin to migrate data from from previous version, and register that plugin with correct release version
+- Each version should have a separate directory, which contains all configuration files and plugins for that version
+- vic-machine upgrade framework will calculate the plugin chain based on old VCH's version and vic-machine version, and then call those plugins sequentially. This will look like r1->r2->r3->f4->latest or r3->r4->r5
+- Run guestinfo validation to figure out if there is any field missing (one possible option here is to translate configuration back to vic-machine input data, and then rerun validator to make sure all vsphere related objects are still exist and correct)
+- Prompt missing parameters if there is any (for new parameters added and cannot be empty)
+- Write new configuration back to appliance guestinfo
+
+With this design, developer is able to add/delete/repurpose existing configuration items, except VCH version. And we'll need one guidline for what to do while one configuration item is changed.
+
+Besides of this solution, we also dicussed another option, that is to revert configuration back to vic-machine create option, and try to recreate a new VCH appliance, and then migrate containers from the old appliance to new appliance. This solution can reduce VCH downtime, and even migrate alive.
+
+We have two concerns for this option. 
+
+First and most important, is that we cannot actually migrate containers from appliance1 to appliance2. This will introduce too many production refactor. For example, image store directory is created uniquely with appliance UUID, network will have conflict, connection over serial port need to be switched, etc.
+
+Second concern is that we'll need to version more objects and logics, includes appliance guestinfo, vic-machine create options, and the logic to revert guestinfo back to vic-machine create options, those are all version sensitive.
+
+We didn't try to figure out how to do it at this time, cause it looks overdesigned.
+
+Note: extraconfig package should always be backward compatible. If it breaks this assumption, upgrade does not work for both options.
+
+### Secret Configuration
+
+In appliance configuration, there are few items are encrypted for security reason, including user password and certificate so far. The encryption key is readable from in guest only. vic-machine is reading configuration through vsphere API, so not be able to decrypt secret information from old appliance configuration.
```

Comment:
Talked to @dougm for the possible vmodl API. 
I tried the API of VirtualMachine.ExtractOvfEnvironment, but it does not return expected key, possibly because we didn't use the XML schema where ovf inject relies on.
So we still do not have solution to read back encryption key even in VC.